### PR TITLE
FFmpeg: Implement image thumbnails

### PR DESCRIPTION
**Pull request by fritsch**
This brings the visionary PR of @ace20022 another step further. With this in place we can basically kick all other image dependencies we have and just use the ffmpeg one. 

Happy for a thourough review.

### CODE REVIEW

File: xbmc/guilib/imagefactory.cpp [hunk 1]
```diff
@@ -43,17 +43,12 @@ IImage* ImageFactory::CreateLoader(const CURL& url)
...
-  else if (strMimeType == "image/gif")
+  if (strMimeType == "image/gif")
     return new Gif();
 #endif//HAS_GIFLIB
 
-  if (strMimeType == "image/webp")
-    return new CFFmpegImage();
-
-  return new CXImage(strMimeType);
+return new CFFmpegImage();
```
Comment (stefansaraev) [inline xbmc/guilib/imagefactory.cpp]: spacing.

**Comment by MrMC** [comment]
very nice, die cxImage die.

**Comment by fritsch** [comment]
We are not done yet - sadly this mjpeg encoder only eats YUVJ420P which is deprecated ... so I am foucking arround with sws_scale to alter the range of YUV420P

**Comment by MrMC** [comment]
any speed differences + or - ?

**Comment by fritsch** [comment]
Just implemented for now - I did not make any tests yet concerning speed.

**Comment by fritsch** [comment]
Original (463 KB):
sws_original

SWS_Rescaled (correct color set - last patch, 281 KB):
sws_right

SWS_Rescaled (wrong color set - without the last patch, 253 KB):
sws_wrong_range

**Comment by ace20022** [comment]
I skimmed through it, nice work! :+1:

**Comment by fritsch** [comment]
@ace20022 yeah - thx for starting it :-)

**Comment by stefansaraev** [comment]
I am ready to clean up after this is in. the question is: do we want to keep CJpegIO. as it has nothing to do with cximage, and we can not drop libjpeg dependency anyway.

**Comment by fritsch** [comment]
@stefansaraev we still need to do some benchmarking - quality wise, file size wise and speed wise. I think I'd like to keep CXImage in there for the first time (but deactivate it)

**Comment by stefansaraev** [comment]
my question is for cjpegio. not cximage ;)

**Comment by ace20022** [comment]
Agree with @fritsch, jpegs are handled different iirc, because cjpegio is ways faster, at least there's a comment somewhere. So we need extensive tests on all platforms.

**Comment by fritsch** [comment]
I am pretty new into this "Image stuff" just came into this while emailing with @ace20022 two days ago - others more into the code need to tell us.

**Comment by stefansaraev** [comment]
sure @ace20022 I am for keeping cjpegio and evenualy removing cximage. but it's up to you guys.

**Comment by MrMC** [comment]
since master is undergoing major construction, nuke out cximage, keep jpgio/gifio and use ffmpeg for everything else. if you keep cximage and fallback to it, you will never find/fix issues.

**Comment by fritsch** [comment]
@davilla:  can you provide some numbers with: http://sprunge.us/WEKV applied?

Here are some results from my testing rigs (all debug build): http://sprunge.us/QEWc

Computing Time: 30.426772 ms Width: 1920 Height: 1080
Computing Time: 0.924089 ms Width: 176 Height: 168
Computing Time: 10.168077 ms Width: 720 Height: 720

**Comment by fritsch** [comment]
JpegIO: http://sprunge.us/VWDT

Computing Time: 28.715415 ms Width: 1920 Height: 1080
Computing Time: 11.248714 ms Width: 1280 Height: 720
Computing Time: 0.454397 ms Width: 168 Height: 168

Not that bad at all :-)

**Comment by fritsch** [comment]
JpegIO Image (352 KB):
jpegio_bench

It's 1.5 times bigger than ours.

**Comment by notspiff** [comment]
why do resolutions differ that much ?

**Comment by fritsch** [comment]
@notspiff I just remove all Thumbnails - then I got into my video directory and scroll slowly through it. Afterwards I do that with music - recently added albums.

I need to plot those - to see a nice statistics, will do (time per w x h) or something

**Comment by notspiff** [comment]
yes, point was you copied 1280 x 720 for jpegio, and 720x720 for ffmpeg, skewing the results somewhat.

**Comment by popcornmix** [comment]
I believe there are two typos in the results (and a change of order) which makes them hard to compare.

**Comment by fritsch** [comment]
Here an image:
comparisonswsjpegio

**Comment by popcornmix** [comment]
I'm assuming the smaller jpeg size from ffmpeg is due to a lower quantiser quality?
If so it's probably worth making the two comparable in output file size before studying the numbers, as the larger sized jpeg will likely be slower to generate.

It's a different question whether a lower quantiser quality is acceptable (it may well be).

**Comment by fritsch** [comment]
I think it's the other way round - more compression should need more time - as you need to compute more ... but yes before handling with figures one should use a common base.

**Comment by popcornmix** [comment]
I think smaller files will be quicker. It is just discarding higher frequency coefficients. The main cost will be in the disk I/O.

**Comment by notspiff** [comment]
what do you plan to do with the formats not supported by ffmpeg? perhaps most notable the raw formats (libdcraw stuff)

**Comment by fritsch** [comment]
@notspiff: Implement them upstream during christmas holidays :-) (just kidding, as said above - no plan yet.

@popcornmix http://www.fastcompression.com/products/jpeg/benchmarks-cuda-jpeg.htm <- does not look like it - but might be special case.

**Comment by stefansaraev** [comment]
imagedecoder addons (like audiodecoders) ?

/me :runner:

**Comment by fritsch** [comment]
@stefansaraev yeah this is one way to go - but I did not propose it - as always: someone needs to do it ... and timewise it looks blacker than black currently for me :-)

**Comment by popcornmix** [comment]
@fritsch The units are GB/s, so I think it is showing smaller files are quicker.

**Comment by fritsch** [comment]
@popcornmix you are right - obviously :-) - now it's time to retire and meet my ergometer :-)

**Comment by ace20022** [comment]
> what do you plan to do with the formats not supported by ffmpeg? perhaps most notable the raw formats (libdcraw stuff)

I've added libraw years ago when I was dumber. The branch is still there.

**Comment by notspiff** [comment]
@stefansaraev initial draft; https://github.com/notspiff/kodi-cmake/commit/7dd3088aded23d84b22dcedc6bffc8911316d9ad

jpeg-turbo seems to work fine(tm); https://github.com/notspiff/imageencoder.jpeg-turbo

File: xbmc/guilib/FFmpegImage.cpp [hunk 2]
```diff
@@ -309,9 +334,174 @@ bool CFFmpegImage::CreateThumbnailFromSurface(unsigned char* bufferin, unsigned
                                              unsigned char* &bufferout,
                                              unsigned int &bufferoutSize)
 {
-  return false;
+  // It seems XB_FMT_A8R8G8B8 mean RGBA and not ARGB
+  if (format != XB_FMT_A8R8G8B8)
+  {
+    CLog::Log(LOGERROR, "Supplied format: %d is not supported.", format);
```
Comment (Jalle19) [inline xbmc/guilib/FFmpegImage.cpp]: return false?

**Comment by iLLiac4** [comment]
Hi. Is it possible that this commit breaks the transparent icons in gui?

File: xbmc/guilib/FFmpegImage.cpp (hunk 2 above)
Comment (fritsch) [inline xbmc/guilib/FFmpegImage.cpp]: Thx - yes of course!

**Comment by fritsch** [comment]
@iLLiac4 this is not even merged yet ...

File: xbmc/guilib/FFmpegImage.cpp [hunk 3]
```diff
@@ -309,9 +334,175 @@ bool CFFmpegImage::CreateThumbnailFromSurface(unsigned char* bufferin, unsigned
                                              unsigned char* &bufferout,
                                              unsigned int &bufferoutSize)
 {
-  return false;
+  // It seems XB_FMT_A8R8G8B8 mean RGBA and not ARGB
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: I think it's becuase of https://ffmpeg.org/doxygen/2.8/pixfmt_8h.html#a9a8e335cf3be472042bc9f0cf80cd4c5

**Comment by iLLiac4** [comment]
I know it is not but it is included in this test build http://forum.kodi.tv/showthread.php?tid=250817&pid=2187006#pid2187006 so i assume that this was the cause for the mentioned behavior.

File: xbmc/guilib/FFmpegImage.cpp [hunk 4]
```diff
@@ -266,7 +291,7 @@ bool CFFmpegImage::Decode(unsigned char * const pixels, unsigned int width, unsi
     width, height, AV_PIX_FMT_RGB32, SWS_BICUBIC, NULL, NULL, NULL);
 
   sws_scale(context, m_pFrame->data, m_pFrame->linesize, 0, m_height,
-    pictureRGB->data, pictureRGB->linesize);
+      pictureRGB->data, pictureRGB->linesize);
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: Just in case you change something else as well: either pad it to context or leave it as is

**Comment by notspiff** [comment]
and with https://github.com/notspiff/kodi-cmake/commit/08ce46ba0e1b351e56ca437ce681d6668ff15648(rewritten commit) https://github.com/notspiff/imageencoder.cximage also works.

**Comment by MilhouseVH** [comment]
> Hi. Is it possible that this commit breaks the transparent icons in gui?

It would appear that png's are being cached as jpg's.

**Comment by notspiff** [comment]
yes that is the case, because ffmpegimage always outputs a jpeg.

**Comment by fritsch** [comment]
thx @notspiff - so which way forward do you suggest?

How is it decided until now if png is to be cached? Do we check the A channel for non 255?

**Comment by akva2** [comment]
@ace20022 since you know this stuff better than me: can you give a abstract interface used between Gif and TextureManager, with the aim to extend iimage ?

**Comment by akva2** [comment]
nah, it's decided through keeping the format. png -> png, jpeg -> jpeg.

**Comment by akva2** [comment]
i think the addons is the way forward.

**Comment by fritsch** [comment]
Okay - so checking the file ending with StringUtils? and using png instead?

**Comment by akva2** [comment]
well, it's done internally in CXImage class right now. the mime type is passed to the constructor and it will write with the same format.

**Comment by fritsch** [comment]
I think I am blind - from the interface I only see: https://github.com/fritsch/xbmc/blob/9d91a763c1a3e6c6dec93cfc7850ac4e990f9ed3/xbmc/guilib/FFmpegImage.cpp#L333 and here I can check if it's .png or .jpg and change the encoder instead.

**Comment by akva2** [comment]
yes, it's not part of the interface, just the cximage constructor.

**Comment by popcornmix** [comment]
I thought the cache to png/jpeg was done based on alpha:
https://github.com/xbmc/xbmc/blob/master/xbmc/TextureCacheJob.cpp#L109
(or at least the filename extension depends on presence of alpha)

**Comment by fritsch** [comment]
@popcornmix yes - that's what I also assumed ... so yes. Let me implement png encoder tonight, too for the case the destFile ends with .png.

**Comment by akva2** [comment]
you are not wrong, but you need to dig deeper to see what happens beneath the surface (see original imagefactory).

**Comment by fritsch** [comment]
The png part is not much code ... as it can cope with RGB32 directly ...

**Comment by ace20022** [comment]
> @ace20022 since you know this stuff better than me: can you give a abstract interface used between Gif and TextureManager, with the aim to extend iimage ?

Please elaborate

**Comment by akva2** [comment]
right now Gif has additional interfaces outside iimage. this makes it impossible to move to an addon. i need a definition of the additional interface required. in particular the multiframe stuff.

**Comment by fritsch** [comment]
Added png encoder based on the file ending

**Comment by MrMC** [comment]
splash chicken vs egg if image decoders move to addons. just a thought.

**Comment by stefansaraev** [comment]
jpegio should stay. gif stays. splash can be jpeg or gif or whatever is supported by ffmpeg.

**Comment by fritsch** [comment]
Jep ... we have a big problem with transparency anyways ... the addon course is also my favorite ... I will remove the "make it default" commit and keep it here for future reference.

**Comment by notspiff** [comment]
ffmpeg supports jpeg, png, whatever for decode. the addons are to add other formats and to allow for using more optimzed ones (e.g. jpegio).

**Comment by fritsch** [comment]
@ace20022 was there a reason you force m_hasAlpha to false all the time? With it set to m_hasAlpha transparent works as they should?

**Comment by MilhouseVH** [comment]
This latest update appears to be working, correctly caching jpg as jpg, and png as png (with alpha).

**Comment by ace20022** [comment]
>@ace20022 was there a reason you force m_hasAlpha to false all the time? With it set to m_hasAlpha transparent works as they should?

I really can't remember, must be a mistake. But shouldn't it depend on the input? Iirc texturecachejob and others rely on that

**Comment by fritsch** [comment]
@ace20022 https://github.com/fritsch/xbmc/commit/c4a50455dd8578ccfcdc822104020516645ad878 <- yes it should depend on the input.

**Comment by MilhouseVH** [comment]
The latest version of this PR is no longer working correctly.

pngs are being cached as jpgs again:
```
22:35:35 134.821671 T:1770390432    INFO:   msg: JpgParse: Not a JPEG file nfs://192.168.0.3/mnt/share/media/Pictures/Zombieland (2009)[BDRip]-logo.png
```

and jpgs are also now failing to be cached:
```
22:33:46  25.800276 T:1770390432   DEBUG: LoadFromFileInternal - Load of nfs://192.168.0.3/mnt/share/media/Video/MoviesSD/Night at the Museum Collection/Night at the Museum 3 (2014)[DVDRip]-poster.jpg failed.
```

It was working - the last two squashes (starting with the code reshuffle) have introduced a change.

**Comment by notspiff** [comment]
@ace20022 converted your raw stuff to an addon https://github.com/notspiff/imageencoder.raw (plus some fixes).

**Comment by Razzeee** [comment]
@notspiff I feel stupid for asking, but shouldn't that be a decoder?

**Comment by akva2** [comment]
@razzeee the iimage interface defines 3 functions, 2 which are related to decoding and 1 which is related to encoding. it seems a bit much to split this in two extension points so i went with encoder, since that entails "more" than decoding. as far as the addon id is concerned, we can do whatever of course.

**Comment by Razzeee** [comment]
Thanks for clarification!

**Comment by fritsch** [comment]
@MilhouseVH yes - it was. The last fixup I did was too late ... and I should have been in bed already ... should work again now.

**Comment by MilhouseVH** [comment]
@popcornmix: this PR is a problem on RPi when https://github.com/fritsch/xbmc/commit/5395829ac497034b1586f1a290c54dded035fcba is included.

It seems the fallback method isn't working for jpgs.

If https://github.com/fritsch/xbmc/commit/5395829ac497034b1586f1a290c54dded035fcba is reverted then jpgs work as expected on RPi, although perhaps using CXImage rather than FFmpeg, so not really achieving the aim of this PR.

But when https://github.com/fritsch/xbmc/commit/5395829ac497034b1586f1a290c54dded035fcba is included, any progressive jpegs fail to load.

Any ideas why that might be?

**Comment by popcornmix** [comment]
Is there an error in log when progressive jpeg fails?

**Comment by MilhouseVH** [comment]
There is no error, the images simply fail to load.

This is a log (debug enabled towards the end): http://sprunge.us/PWaX

Here is a another log with https://github.com/fritsch/xbmc/commit/5395829ac497034b1586f1a290c54dded035fcba reverted: http://sprunge.us/RODM

**Comment by popcornmix** [comment]
It's not related to the Pi''s GPU acceleration if you disable that (http://paste.ubuntu.com/14054253/)
you'll find that both progressive and non-progressive jpegs fail.

av_probe_input_buffer2 appears to be given a plausible looking jpeg buffer but fails to find a format. I'm guessing this is an ffmpeg configuration issue? Is it being configured to support jpeg decode?

**Comment by popcornmix** [comment]
I noticed when stepping through the code that if loading fails we try again, but seem to be trying exactly the same thing:
https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/Texture.cpp#L271
https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/Texture.cpp#L296

as the loader and fallback loader are both ffmpeg. I wonder if that can be avoided.

**Comment by stefansaraev** [comment]
> as the loader and fallback loader are both ffmpeg. I wonder if that can be avoided.

yes. nuke fallbackloader completely

**Comment by MilhouseVH** [comment]
> Is it being configured to support jpeg decode?

Decoders are left at default (ie. enabled) Encoders are all disabled, and then explicitily enabled (in this case, mjpeg and png encoders are both enabled).

According to `ffmpeg -decoders | grep jpeg`, the following jpeg decoders are available:
```
Decoders:
 V..... = Video
 A..... = Audio
 S..... = Subtitle
 .F.... = Frame-level multithreading
 ..S... = Slice-level multithreading
 ...X.. = Codec is experimental
 ....B. = Supports draw_horiz_band
 .....D = Supports direct rendering method 1
 ------
 VF...D jpeg2000             JPEG 2000
 V....D jpegls               JPEG-LS
 V....D mjpeg                MJPEG (Motion JPEG)
 V....D mjpegb               Apple MJPEG-B
 V..... smvjpeg              SMV JPEG
 A....D adpcm_ima_smjpeg     ADPCM IMA Loki SDL MJPEG
```
So the mjpeg decoder is already enabled.

**Comment by popcornmix** [comment]
Does --enable-encoder=png need to go in https://github.com/xbmc/xbmc/blob/master/tools/depends/target/ffmpeg/Makefile#L22 for universal builds?

**Comment by fritsch** [comment]
The last commit is brute force ... but it seems ffmpeg is not that smart when guessing some jpeg from memory :-(

@FernetMenta any better idea to get it probing more thoroughly?

**Comment by MilhouseVH** [comment]
OK this is looking good now - all images are being cached (jpg and png), with no failures.

Debug log: http://sprunge.us/RTaK

Note the "CFFmpegImage I was really used here!" message being logged is a [personal addition](http://sprunge.us/eNYW) and not part of this PR. There are no entries from "CXImage".

**Comment by popcornmix** [comment]
@fritsch I'd say you want to specify jpeg explicitly.
Stepping through the ffmpeg probe code gives the jpeg buffer to every supported codec (and there are loads). It will be a lot quicker to skip probing if at all possible.

**Comment by MilhouseVH** [comment]
Although there is an issue with the aspect ratio of some posters and fanart:

Note the leftmost "Mockingjay" poster in the "Recently added movies" widget:

CFFmpegImage (with this PR - incorrect proportiions):
s1

CXImage (without this PR - correct proportiions):
s2

Fanart:

CFFmpegImage (with this PR - incorrect proportiions):
s3

CXImage (without this PR - correct proportiions):
s4

The "Mockingjay" poster is a little unusual in that it is 3000h  x 2000w (most of my posters are 1500h x 1000w). Oddly, it is displayed correctly in the "Recently added" view, but incorrectly in the widget.

Most of my fanart is 1920w x 1080h, however the Jurassic World fanart that is being scaled incorrectly is 3840w x 2160 h. Another 3840w x 2160h fanart shows similarly incorrect scaling/stretching, so this problem only seems to affect "oversized" artwork.

Test device is an RPi/RPi2 with 1920x1080 display. My advancedsettings.xml include:
```
  <fanartres>1080</fanartres>
  <imageres>720</imageres>
 ```

With the above settings, CFFmpegImage is caching the Jurassic World fanart (original size: 3840w x 2160h) at a size of 720w x 720h, while other 1920w x 1080h fanart is being correctly scaled and cached at 1920w x 1080h.

Similarly, the Mockingjay poster (original size: 3000h x 2000w) is being cached by CFFmpegImage at 703w x 720h, whereas CXImage would cache this poster with proportions of 480w x 720h.

I can provide the original artwork on request.

**Comment by MilhouseVH** [comment]
In addition, after reviewing the debug log (http://sprunge.us/RTaK), there are some bogus details being reported, eg.:
```
23:48:26 193.016052 T:1582298016   DEBUG: COMXImageFile::ReadFile nfs://192.168.0.3/mnt/share/media/Video-Private/MoviesHD/Lucy (2014)[BDRip]-fanart.jpg GetCodingType=0x7fffffff (1920x438)
```
when the original artwork and that written into the cache are both 1920x1080 - I've no idea where the 1920x438 is coming from.

**Comment by fritsch** [comment]
@popcornmix can do yes - same for png?

**Comment by fritsch** [comment]
@MilhouseVH yeah - those are too large for your surfaces ... so we use the maximum in one dimension we can get.

**Comment by fritsch** [comment]
@MilhouseVH can you please retest - the wrong ratio I did not fix yet - but will do. For some known formats I now ask for the relevant pipes directly.

**Comment by popcornmix** [comment]
@fritsch I think you should always set the image format explicitly if you know it. Probing will be slower (and is more likely to go wrong). So yes to PNG.

**Comment by fritsch** [comment]
@popcornmix can you check: https://github.com/fritsch/xbmc/commit/1d38509b680b5a6020abc3dffc7a5d5ccc7c4ac2 i am highly confused with that find method ... documentation is not that good, people on IRC don't answer yet - do you think it's correct to use the pipe names?

**Comment by popcornmix** [comment]
@fritsch : the probe functions don't seem to be called and it seems to work, so I guess it's okay.

**Comment by fritsch** [comment]
jenkins build this please dooday dooday day

**Comment by fritsch** [comment]
just asking jenkins to find some errors, will squash most of the commits and then keep it standing here - we need to decide if cximage is removed right away and we actually find out which user formats (that are actually in use) don't work anymore. For those the binary addon approach by @notspiff would be the solution.

I want to get this  tested as soon as possible. Cause we would not realize any issues if we kept cximage as fallback.

**Comment by fritsch** [comment]
jenkins build this please

File: xbmc/guilib/FFmpegImage.cpp [hunk 5]
```diff
@@ -41,10 +42,33 @@ struct MemBuffer
...
+    avOutctx = nullptr;
+    sws_freeContext(sws);
+    sws = nullptr;
+  }
+};
 static size_t Clamp(int64_t newPosition, size_t bufferSize, size_t offset)
 {
   size_t clampedPosition = 0;
-  if (newPosition < 0 || offset + newPosition < 0)
+  if (newPosition < 0)
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: Imo the whole clamp is wrong, drop offset and check for <0 and >= bufferSize

Comment (fritsch) [inline xbmc/guilib/FFmpegImage.cpp]: https://github.com/fritsch/xbmc/commit/3900a8ead92cea923e29918912d5e732a6674c86 <- did you think that way?

Comment (fritsch) [inline xbmc/guilib/FFmpegImage.cpp]: Or a one liner:
```
return std::min(std::max((int64_t) 0, newPosition), (int64_t) (bufferSize -1));
```

**Comment by MilhouseVH** [comment]
Any thoughts on this [forum report](http://forum.kodi.tv/showthread.php?tid=250817&pid=2192159#pid2192159)?

**Comment by fritsch** [comment]
All working for me. I replied there.

2015-12-21 2:53 GMT+01:00 MilhouseVH <notifications@github.com>:

> Any thoughts on this forum report
> <http://forum.kodi.tv/showthread.php?tid=250817&pid=2192159#pid2192159>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8583#issuecomment-166172998>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by fritsch** [comment]
jenkins build this please

**Comment by fritsch** [comment]
@wsnipex @MartijnKaijser @popcornmix @MilhouseVH I plan to push this - a cximage removal PR will follow two weeks later when we got more feedback on this new default image loader. @stefansaraev has already done that.

@ace20022 and I decided that we want to have raw support in core, therefore we will transform @notspiff's code of his addon image solution to kodi during the next weeks. But we also want to have this new possibility to add binary addons for images for all formats we don't support anymore now.

This code is tested since some days / weeks in the famouse Milhouse builds and for now it is looking really promissing.

@wsnipex is there anything I have forgotten concerning the depends? Should we add something into a readme, e.g. we need encoder mjpeg, png now?

**Comment by akva2** [comment]
my .02; putting anything but ffmpeg in core is pure stupid. absolute 100% refined stupid,

**Comment by fritsch** [comment]
@akva2 if we can solve the scaling issue, e.g. providing an ffmpeg scaler for image addons and if we can take care that such addons are available for all platforms without issues - then I am the share your opinion. From my last look at it - it was quite hard to get that done right, but if you say: hey, we get that going and it's easy to provide the addons for all platforms at the end. I am also fine with me.

My point in general was: If we want to support raw, we need to take care that it is easily doable "out of the box".

**Comment by akva2** [comment]
sure. if anything the current addons prove that the iimage interface is not enough. but we can change interfaces. bringing stuff into core is much harder to undo.

it's not that hard to do that afaict.. we either do it outside the iimage interface, leaving it up to the app (ie addons always output unscaled images), or we add a callback class and let the addons handle it by using that.

**Comment by fritsch** [comment]
Alternatively: Before FFmpegImage is contstructed, we ask given image addons for their supported mime types and use those to open the files, afterwards we pass the result (png raw stream, something else) to the already given infrastructure?

**Comment by akva2** [comment]
uhm, that's what it does now.

**Comment by akva2** [comment]
i dont see how that changes anything in regards to scaling.

**Comment by fritsch** [comment]
Then all fine ... I only read the addon code for libraw, sorry for that. So yes - then we have a solution for the kodi clue part. Sorry!

**Comment by fritsch** [comment]
FFmpeg's Decode method has a scaling method: https://github.com/fritsch/xbmc/commit/c9ed88e26165e45884340abe0e01c4d4660314c4 so whenever the given texture targes are too small images are rescaled, so if it's going through that, they are rescaled. Though there is a little overhead (one additional full copy) that could be solved by static ImageScaler, that btw. FFmpegImage could then reuse again.

**Comment by ace20022** [comment]
>ImageScaler

That's what I had in mind.This could also be an addon, probably not, or?

**Comment by fritsch** [comment]
Yes. We also have encoders that user internal ffmpeg contexts. So an addon depending on sws_scale is doable. Lifetime of buffers, especially resulting buffers need to be discussed but I think @notspiff already has thought through a possible poc that would help us in that regard.

**Comment by wsnipex** [comment]
looks fine depends wise. Not sure if ffmpeg needs external libs for mjpeg or png, but we have at least libpng available. If it needs libpng, then you should add it to ffmpeg depends in https://github.com/xbmc/xbmc/blob/master/tools/depends/target/Makefile#L103

**Comment by fritsch** [comment]
@wsnipex thx for looking, I think it is not needed: https://ffmpeg.org/general.html#Image-Formats it does it internally.

**Comment by fritsch** [comment]
jenkins build this please

File: tools/depends/target/ffmpeg/autobuild.sh [hunk 6]
```diff
@@ -153,6 +153,8 @@ CFLAGS="$CFLAGS" CXXFLAGS="$CXXFLAGS" LDFLAGS="$LDFLAGS" \
 	--enable-libvorbis \
 	--enable-muxer=ogg \
 	--enable-encoder=libvorbis \
+        --enable-encoder=png \
```
Comment (FernetMenta) [inline tools/depends/target/ffmpeg/autobuild.sh]: this is a script and tabs are used.

Comment (fritsch) [inline tools/depends/target/ffmpeg/autobuild.sh]: Jenkins was happy - but I will correct obviously. And bash scripts are tab / space unaware ...

**Comment by Memphiz** [comment]
mhhh jpegio was nuked now and we still link kodi against libjpeg for darwin here:

https://github.com/xbmc/xbmc/blob/master/tools/darwin/Configurations/App.xcconfig.in#L28

What am i missing?

**Comment by fritsch** [comment]
@Memphiz this PR did not nuke it, though @stefansaraev do you have an idea?

**Comment by stefansaraev** [comment]
huh. I missed that as jenkins built all fine. will add to #8772

**Comment by notspiff** [comment]
it wouldn't show up as an error. the library is there (needed by texturepacker) and the linker simply doesn't import any symbols since there are no references.

**Comment by Memphiz** [comment]
Nope texturepacker is native :D 
it built fine because we still build libjpeg-turbo for target in depends (for pythonmodule-pil needs it) - we can drop the link in kodi then - thx for taking care of it @stefansaraev - not sure if other platforms link libjpeg into kodi aswell though ...

**Comment by notspiff** [comment]
true that. my point is still valid though ;)

other platforms have been pruned (configure does things properly and have no hard coded link list afaik).

**Pull request merged**